src/plccng/spec/syntax/validations/ll1/check_left_recursion.py

```diff
@@ -0,0 +1,67 @@
+from .Grammar import Grammar
+import copy
+
+def check_left_recursion(grammar: Grammar):
+    return LeftRecursionChecker(grammar).check()
+
+class LeftRecursionChecker:
+    def __init__(self, grammar: Grammar):
+        self.grammar = grammar
+        self.grammarCopy = copy.deepcopy(grammar)
+        self.substitutions = {}
+        self.nonterminalsInOrder = list(self.grammarCopy.getRules().keys())
+        self.rulesDict = self.grammarCopy.getRules()
+
+    def check(self):
+        allLeftRecursionInstances = []
+        for n in self.nonterminalsInOrder:
+            self._handleLeftRecursionSubstitutions(n)
+            for rule in self._findDirectLeftRecursion(n):
+                allLeftRecursionInstances.append(self._traverseSubstitutionsForOriginal(n, rule))
+        return allLeftRecursionInstances if allLeftRecursionInstances else None
+
+    def _handleLeftRecursionSubstitutions(self, n):
+        currNontermRules = self.rulesDict[n]
+        for prevNonterm in self.nonterminalsInOrder[:self.nonterminalsInOrder.index(n)]:
+            self._substitute(currNontermRules, prevNonterm)
+
+    def _substitute(self, currNontermRules, prevNonterm):
```

Comment:
What gets substituted into what? Let's get that into the name of the method and/or the parameter names.